colony_picker/inverse_kinematics.py

```diff
@@ -0,0 +1,634 @@
+                    solved = True
+                    break
+    return (joint_angles, solved)
+
+# ************************************** #
+#     HELPER FUNCTIONS FOR ANIMATIONS    #
+# ************************************** #
+
+
+def draw_coordinate_system(plotter: pv.Plotter, t_mat: np.ndarray,
+                           base: bool = False, name: str = ""):
+    """draws a coordinate system on the plotter with the orientation and
+    position described by t_mat and with the x, y, and z axes denoted by red,
+    green, and blue colors. 
+
+    Args:
+        plotter (pv.Plotter): the plotter object to draw the coordinate system
+            on.
+        t_mat (np.ndarray): a transformation matrix describing the position and
+            orientation of the coordinate system relative to the base reference
+            frame. 
+        base (bool, optional): whether the coordinate system being drawn
+            represents the base of the robot. Defaults to False.
+        name (str, optional): the name of the joint whose coordinate system is
+            being drawn so that the position and orientation of the coordinate
+            system can be updated with user input. Rather than a new coordinate
+            system being drawn, coordinate systems with names can be updated.
+            Defaults to "".
+    """
+    colors = ["red", "green", "blue"]
+    position = t_mat[:3, 3]
+    sphere = pv.Sphere(center=position, radius=4)
+    if base:
+        sphere_color = "black"
+    else:
+        sphere_color = "yellow"
+    plotter.add_mesh(sphere, color=sphere_color, name=(name + "_sphere"))
+    for axis_idx in range(3):
+        axis = pv.Arrow(
+            start=position, direction=t_mat[:3, axis_idx], scale=50)
+        plotter.add_mesh(axis, color=colors[axis_idx], name=(
+            name + f"_coord_{axis_idx}"))
+
+
+def draw_links(plotter: pv.Plotter, joint_positions: np.ndarray):
+    """Draws the backbone of the robot arm on plotter given a set of joint
+    positions relative to the base of the arm. 
+
+    Args:
+        plotter (pv.Plotter): the pyvista plotter object to draw the robot arm
+            on
+        joint_positions (np.ndarray): an n x 3 array denoting the position of 
+            each joint relative to the base of the robot in the format 
+            [[x 1, y 1, z 1], [x 2, y 2, z 2], ... , [x n, y n, z n]] where
+            n = number of joints + 1.
+    """
+    poly = pv.PolyData()
+    poly.points = joint_positions
+    cells = np.full((len(joint_positions)-1, 3), 2, dtype=np.int_)
+    cells[:, 1] = np.arange(0, len(joint_positions)-1, dtype=np.int_)
+    cells[:, 2] = np.arange(1, len(joint_positions), dtype=np.int_)
+    poly.lines = cells
+    poly["scalars"] = np.arange(poly.n_points)
+    tube = poly.tube(radius=10)
+    plotter.add_mesh(tube, color="black", name="robot skeleton", opacity=0.5)
+
+
+def draw_robot_arm(thetas: np.ndarray, p: pv.Plotter, dh_params: np.ndarray):
```

Comment:
Is there a reason that the plotter argument is second, when in other functions it's first, and the argument name is `p` instead of `plotter`?